File: Data_Exploration_of_Cyptocurrency_time_series/SOM.py
import numpy as np
import random
import DataQuantizer

class SOM:

    def __init__(self, data, neuronsWidth, neuronsHeight):
        self.data = data
        dimData = np.size(data,1)
        self.neuronsWidth = neuronsWidth
        self.neuronsHeight = neuronsHeight
        self.neurons = np.zeros((neuronsHeight, neuronsWidth, dimData))
        for i in range(0,np.size(self.neurons,0)):
            for j in range(0,np.size(self.neurons,1)):
                for s in range(0,np.size(self.neurons,2)):
                    self.neurons[i,j,s] = random.random()

        self.dimData = dimData

    def trainOnline(self, eta, radius, epochs):
        neighborhoodCalculator = NeighborhoodCalculator()
        distanceCalculator = DataQuantizer.DistanceCalculator()
        nData = np.size(self.data,0)
        for i in range(0,epochs):
            #Vamos a ir disminuyendo linealmente el radio y la tasa de aprendizaje
            sigma = round(radius*(1 - (i-1)/epochs))
            learningRate = eta*(1 - 1/epochs)
            #Seleccionamos un vector aleatoriamente
            randVector = self.data[random.randint(0,nData - 1),:]
            #Encontramos el BMU
            ind1 = 0
            ind2 = 0
            minDistance = float('Inf')
            for row in range(0,self.neuronsHeight):
                for col in range(0,self.neuronsWidth):
                    dist = distanceCalculator.euclideanDistance(randVector, self.neurons[row,col,:])
                    if(dist < minDistance):
                        minDistance = dist
                        ind1 = row
                        ind2 = col

            indsToModify = neighborhoodCalculator.getInclusiveNeighborhood(ind1, ind2, self.neuronsWidth, self.neuronsHeight, sigma)
            bmuNeuron = self.neurons[ind1,ind2,:]

            for j in range(0,len(indsToModify)):
                inds = indsToModify[j]
                neuron = self.neurons[inds[0],inds[1],:]
                neighborhoodDist = neighborhoodCalculator.getNeighborhoodDistance(bmuNeuron, neuron)
                bmuDist = 1
                if(neighborhoodDist != 0):
                    norm = distanceCalculator.euclideanDistance(neuron, bmuNeuron)
                    bmuDist = np.exp(-(norm**2)/(2*neighborhoodDist**2))

                self.neurons[inds[0],inds[1],:] = self.neurons[inds[0],inds[1],:] - learningRate*bmuDist*(neuron - randVector)

    def getUMatrix(self):
        u = np.zeros((2*self.neuronsHeight - 1, 2*self.neuronsWidth - 1))
        neighborCalc = NeighborhoodCalculator()
        distCalc = DataQuantizer.DistanceCalculator()
        for i in range(0,self.neuronsHeight):
            for j in range(0,self.neuronsWidth):
                neighbors = neighborCalc.getInclusiveNeighborhood(i,j,self.neuronsWidth, self.neuronsHeight, 1)
                value = 0
                neuron = self.neurons[i,j,:]
                for k in range(0,len(neighbors)):
                    inds = neighbors[k]
                    dst = distCalc.euclideanDistance(neuron, self.neurons[inds[0],inds[1]])
                    value = value + (1/len(neighbors))*dst

                    if(inds[0] == i and inds[1] != j):
                        u[int(2*i - 1), int(2*np.min([inds[1],j]))] = dst
                    elif(inds[1] == j and inds[0] != i):
                        u[int(2*np.min([inds[0],i])),int(2*j -1)] = dst
                u[int(2*i - 1), int(2*j - 1)] = dst
        uHeight = np.size(u,0)
        uWidth = np.size(u,1)
        for i in range(0,uHeight):
            for j in range(0,uWidth):
                if(i%2 == 0 and j%2 == 0):
                    neighbors = neighborCalc.getInclusiveNeighborhood(i,j,uWidth, uHeight, 1)
                    s = 0
                    for k in range(0,len(neighbors)):
                        nInds = neighbors[k]
                        s = s + (1/len(neighbors))*u[nInds[0],nInds[1]]
                    u[i,j] = s
        return u

    def findWinner(self, vector):
       row = 0
       col = 0
       minDist = float('Inf')
       distanceCalculator = DataQuantizer.DistanceCalculator()
       for i in range(0,self.neuronsHeight):
           for j in range(0,self.neuronsWidth):
               neuron = self.neurons[i,j,:]
               dst = distanceCalculator.euclideanDistance(vector, neuron)
               if(dst < minDist):
                   row = i
                   col = j
                   minDist = dst
       return [row, col]         


class NeighborhoodCalculator:

    def getInclusiveNeighborhood(self, row, col, gridWidth, gridHeight, radius):
        inds = []
        inds.append([row,col])
        for i in range(0,radius):
            for p in range(0,len(inds)):
                pt = inds[p]
                neighbors = []
                for j in range(-1, 2):
                    for s in range(-1,2):
                        newInds = [pt[0] + j, pt[1] + s]
                        if(not (j == 0 and s== 0) and (j == 0 or s == 0) and newInds[0] >= 0 and newInds[0] < gridHeight and newInds[1] >= 0 and newInds[1] < gridWidth):
                            neighbors.append(newInds)

                for j in range(0,len(neighbors)):
                    neighbor = neighbors[j]

                    if(not neighbors in inds):
                        inds.append(neighbor)
        return inds


    def getNeighborhoodDistance(self, x1, x2):
        dst = 0
        if(len(x1) == len(x2)):
            for i in range(0,len(x1)):
                dst += np.abs(x1[i]-x2[i])
            return dst
        else:
            print("Operación invalida, los arreglos deben ser del mismo tamaño")

File: SOM/SOM.m
classdef SOM 
    
    properties
        data;
        dimData;
        neurons;
        neuronsWidth;
        neuronsHeight;
    end
    
    
    methods
        function obj = SOM(data, neuronsWidth, neuronsHeight)
            obj.data = data;
            dimData = size(data, 2);
            %Inicializamos aleatoriamente las neuronas
            obj.neuronsWidth = neuronsWidth;
            obj.neuronsHeight = neuronsHeight;
            obj.neurons = rand(neuronsHeight, neuronsWidth, dimData);
            obj.dimData = dimData;
        end
        
        function obj = trainOnline(obj, eta, radius, epochs)
            neighborhoodCalculator = NeighborhoodCalculator();
            nData = size(obj.data, 1);
            f = waitbar(0,'Calculando SOM');
            for i = 1:epochs
                %Vamos a ir disminuyendo linealmente el radio y el factor
                %de aprendizaje
                sigma = round(radius*(1 - (i-1)/epochs));
                learningRate = eta*(1 - i/epochs);
                waitbar(i/epochs,f, "Calculando SOM radio "+string(sigma));
                %seleccionamos un dato aleatorio
                randVector = obj.data(randi([1,nData]), :)';
                
                %encontramos el vector w que mas se parezca a ese vector 
                ind1 = 0;
                ind2 = 0;
                minDistance = Inf;
                for row = 1:obj.neuronsHeight
                    for col = 1:obj.neuronsWidth
                        dist = norm(reshape(obj.neurons(row, col, :),[obj.dimData 1]) - randVector);
                        if(dist < minDistance)
                           ind1 = row;
                           ind2 = col;
                           minDistance = dist;
                       end
                    end
                end
                
                %Determinamos los vecinos a modificar
                indsToModify = neighborhoodCalculator.getInclusiveNeighborhoodIndices(ind1, ind2, obj.neuronsWidth, obj.neuronsHeight, sigma);
                bmuNeuron = reshape(obj.neurons(uint32(ind1), uint32(ind2), :),[obj.dimData, 1]);
                %Actualizamos el peso de dichas neuronas
                for j = 1:size(indsToModify, 1)
                    inds = indsToModify(j,:);              
                    neuron = reshape(obj.neurons(uint64(inds(1)), uint64(inds(2)),:),[obj.dimData, 1]);
                    neighborhoodDist = neighborhoodCalculator.getNeighborhoodDistance(bmuNeuron, neuron);
                    if(neighborhoodDist ~= 0)
                        bmuDist = exp(-((norm(neuron - bmuNeuron)^2)/(2*neighborhoodDist^2)));
                    else
                       bmuDist = 1; 
                    end
                    obj.neurons(inds(1), inds(2),:) = neuron - learningRate * bmuDist*(neuron - randVector);                    
                end
            end
            close(f);
        end
        
        
        function u = getUMatrix(obj)
            u = zeros(2*obj.neuronsHeight - 1, 2*obj.neuronsWidth - 1);
            neighborCalc = NeighborhoodCalculator();
            distCalc = DistanceCalculator();
            for i = 1:obj.neuronsHeight
               for j = 1:obj.neuronsWidth
                   neighbors = neighborCalc.getInclusiveNeighborhoodIndices(i,j,obj.neuronsWidth, obj.neuronsHeight, 1);
                   value = 0;
                   neuron = reshape(obj.neurons(i,j,:),[obj.dimData,1]);
                   for k = 1:size(neighbors, 1)
                       inds = neighbors(k,:);
                       dst = norm(neuron -reshape(obj.neurons(inds(1), inds(2),:),[obj.dimData,1]));
                       value = value + (1/size(neighbors,1))*dst;
                       if(inds(1) == i && inds(2) ~= j)
                           u(2*i-1,2*min(inds(2), j)) = dst;
                       elseif(inds(2) == j && inds(1)~= i)
                           u(2*min(inds(1), i), 2*j-1) = dst;
                       end
                       endf
                   u(uint64(2*i-1), uint64(2*j-1)) = value;
               end
            end
            %Hasta aquí, ya habremos llenado una buena parte de los valores
            %de U, pero faltarán por llenar algunos huecos
            uHeight = size(u,1);
            uWidth = size(u,2);
            for i = 1:uHeight
                for j = 1:uWidth
                   if(mod(i,2) == 0 && mod(j,2) == 0)
                       neighbors = neighborCalc.getInclusiveNeighborhoodIndices(i,j,uWidth, uHeight, 1);
                       s = 0;
                       for k = 1:size(neighbors, 1)
                          nInds = neighbors(k,:);
                          s = s + (1/size(neighbors,1))*u(nInds(1), nInds(2));
                       end
                       u(i,j) = s;
                   end
                end 
            end
        end
        
        function class = somClassify(obj, vector)
            class = 0;
            classCounter = 0;
            minDist = Inf;
            for i = 1:obj.neuronsHeight
               for j = 1:obj.neuronsWidth
                   d = norm(reshape(obj.neurons(i, j, :),[obj.dimData 1]) - vector');
                   if(d < minDist)
                      class = classCounter;
                      minDist = d;
                   end
                   classCounter = classCounter + 1;
               end
            end
        end
        
        
        
        
    end
    
    
    
    
    
    
end